### PR TITLE
Replace legacy dropdown component with new Tippy dropdown-button. Fix #7366

### LINKED ISSUE #7366: Page editing actions dropdown doesn’t support keyboard or screen reader navigation
### Issue Summary

Initially reported by @jessemenn. On the page editing UI, the actions dropdown is very brittle with keyboard and screen reader navigation:

- The dropdown’s toggle button isn’t focusable
- Items within the dropdown are focusable even when hidden.

### Steps to Reproduce

1. Go to the page editing UI on any Wagtail site
2. Reach the "Save draft" button
3. Try to reach any action within the dropdown

---

* I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: yes

### Technical details

Tested in Chrome macOS & Safari VoiceOver. I’d expect this to be an issue for all browsers, all screen readers, in all Wagtail releases over the last few years at least.

### PR DESCRIPTION
**Pull request by thibaudcolas**
Fixes #7366. Goes some of the way towards #9030. Replaces #11084. This converts the "dropdown legacy" component to Tippy.js like all other dropdowns and tooltips in Wagtail, and also refactors the "footer actions" templates for simplicity. I also made a few related changes.

This new "button with dropdown next to it with buttons inside" looks very different from other dropdowns / tooltips, so I added theming support to our `DropdownController`. So this is easier to reuse in the future, I’ve also switched it to a template tag rather than copy-pasting the markup everywhere.

This component is used for:

- Pages and snippets’ "footer actions"
- Workflows and workflow tasks’ footer actions
- CSV/spreadsheet export

In addition, the footer styles I changed here are used for:

- Bulk actions in listings
- Inspect views
- Pages and snippets‘ editing forms
- Workflow and settings editing forms

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
-   ~~[ ] For Python changes: Have you added tests to cover the new/fixed behaviour?~~
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [x] **Please list the exact browser and operating system versions you tested**: Chrome 121, Firefox 121, Safari 17 on macOS 14.0
    -   [x] **Please list which assistive technologies [^3] you tested**: Keyboard, WHCM, VoiceOver, Voice Control
-   ~~[ ] For new features: Has the documentation been updated accordingly?~~

There were a few specific scenarios that are easy to miss so I took screenshots:

## RTL support

rtl-dropdown-buttons

It’s important the button mirrors, including which corners are rounded, regardless of the orientation of the dropdown.

## Multilingual support

I didn’t got exactly for the intended designs but did reduce the width of the footer actions – while make sure the text doesn’t overflow more than it used to (tested in German).

dropdown-mutlilingual

dropdown overflow

## Workflows without enable or disable permissions

Without the permissions, the workflow and workflow task dropdowns would be empty. This was a pre-existing issue, which doesn’t break the component, but I chose to fix it. Looks like this:

workflow-save-only

## Revert views for pages and snippets

… where the actions are warning-themed:

warning-themed

### CODE REVIEW

File: client/scss/components/_bulk_actions.scss [hunk 1]
```diff
@@ -45,6 +45,9 @@
     }
 
     .bulk-actions-buttons {
+      @include unlistimmediate();
+      display: flex;
+      gap: theme('spacing.4');
```
Comment (thibaudcolas) [inline client/scss/components/_bulk_actions.scss:50]: Replicates styles as-is, previously done for "all lists in all footers".

File: client/scss/components/_dropdown-button.scss [hunk 2]
```diff
@@ -0,0 +1,136 @@
+// stylelint-disable selector-attribute-name-disallowed-list
+$separator: 1px solid theme('colors.white-15');
+$radius: theme('borderRadius.sm');
+
+.w-dropdown-button {
+  // Cosmetic details based on live tooltip placement,
+  // implemented CSS-only by using Tippy.js `data-placement` and `:has`.
```
Comment (thibaudcolas) [inline client/scss/components/_dropdown-button.scss:7]: This is really hard to follow unless you’re atuned with logical corner radius properties. The RTL gif in the README shows that this works.

File: client/scss/components/_dropdown-button.scss [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  display: flex;
+  // Make sure the tooltip within will match this element’s width.
+  position: relative;
+
+  [data-tippy-root] {
+    // Make sure the tooltip within will match this element’s width.
+    width: 100%;
+  }
+
+  @supports not selector(:has(*)) {
```
Comment (thibaudcolas) [inline client/scss/components/_dropdown-button.scss:29]: As of now, Firefox 115 ESR is the only browser we support which doesn’t have `:has` support.

File: client/scss/components/_dropdown-button.scss [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+  &:last-child {
+    border-bottom: var(--last-item-border-bottom);
+    border-end-start-radius: var(--last-item-end-start-radius);
+    border-end-end-radius: var(--last-item-end-end-radius);
+  }
+}
+
+.w-dropdown-button > .warning {
+  + .w-dropdown .button {
```
Comment (thibaudcolas) [inline client/scss/components/_dropdown-button.scss:121]: Bit of a mad selector here but this essentially means we don’t have to change where that `warning` class is applied too much. I suspect we’d want to remove those yellow styles as part of completing #9030 anyway, so I didn’t spend too much time assessing this.

File: client/scss/components/_dropdown-button.scss [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    border-inline-end-color: theme('colors.warning.100');
+    border-top-color: theme('colors.warning.100');
+    border-bottom-color: theme('colors.warning.100');
+  }
+}
+
+// Reduces the visibility of secondary actions like "Unpublish" or "Delete".
+.w-dropdown-button .action-secondary {
+  color: theme('colors.text-label-menus-default');
+}
```
Comment (thibaudcolas) [inline client/scss/components/_dropdown-button.scss:136]: Also left for #9030. Not too sure why we do this.

File: client/scss/components/_dropdown.scss [hunk 6]
```diff
@@ -21,6 +21,10 @@
     @apply w-w-4 w-h-4 w-mr-3 w-transition w-opacity-50 w-shrink-0;
   }
 
+  .icon-wrapper .icon {
+    @apply w-mr-0;
+  }
```
Comment (thibaudcolas) [inline client/scss/components/_dropdown.scss:26]: This is necessary for bicolor buttons within dropdowns. See for example redirects’ header export actions.

File: client/scss/components/_footer.scss [hunk 7]
```diff
@@ -1,189 +1,112 @@
 @use 'sass:math';
 
 .footer {
```
Comment (thibaudcolas) [inline client/scss/components/_footer.scss:3]: I’ve done a fair bit of refactoring here – not necessarily changing the styles too much but at least making sure to reduce selector specificity where possible. A lot of styles just moved further down the file so this is hard to review as a line-by-line diff.

File: wagtail/admin/templates/wagtailadmin/workflows/edit.html [hunk 8]
```diff
@@ -62,46 +62,45 @@
...
+                                class="button action-save button-longrunning"
+                                data-controller="w-progress"
+                                data-action="w-progress#activate"
+                                data-w-progress-active-value="{% trans 'Saving…' %}"
+                            >
+                                {% icon name="spinner" %}
+                                <em data-w-progress-target="label">{% trans 'Save' %}</em>
+                            </button>
+                        {% endfragment %}
+                        {% if can_enable or can_disable %}
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/workflows/edit.html:80]: Departure from previous implementation which rendered the dropdown anyway.

File: client/src/controllers/DropdownController.ts [hunk 9]
```diff
@@ -173,4 +193,14 @@ export class DropdownController extends Controller<HTMLElement> {
       rotateToggleIcon,
     ].concat(this.hideOnClickValue ? [hideTooltipOnClickInside] : []);
   }
+
+  /**
+   * Use a different reference element depending on the theme.
+   */
+  getReference() {
```
Comment (lb-) [inline client/src/controllers/DropdownController.ts:200]:
```suggestion
  get reference() {
```

File: client/src/controllers/DropdownController.ts [hunk 10]
```diff
@@ -144,11 +163,12 @@ export class DropdownController extends Controller<HTMLElement> {
       ...(this.hasContentTarget
         ? { content: this.contentTarget as Content }
         : {}),
+      ...themeOptions[this.themeValue],
       trigger: 'click',
       interactive: true,
-      theme: 'dropdown',
       ...(this.hasOffsetValue && { offset: this.offsetValue }),
-      placement: 'bottom',
+      getReferenceClientRect: () => this.getReference().getBoundingClientRect(),
```
Comment (lb-) [inline client/src/controllers/DropdownController.ts:170]:
```suggestion
      getReferenceClientRect: () => this.reference.getBoundingClientRect(),
```

**Review by lb- — COMMENTED** [review]
Super keen to see this in - amazing work here @thibaudcolas .

One small suggestion, not critical. I am happy to approve this but will let @laymonage to a review as was requested.

File: wagtail/admin/templates/wagtailadmin/workflows/edit_task.html [hunk 11]
```diff
@@ -51,22 +51,28 @@
...
+                                        type="submit"
+                                    >
+                                        {{ view.enable_item_label }}
+                                    </button>
+                                {% elif can_disable %}
+                                    <a href="{{ view.get_delete_url }}" class="button">{{ view.delete_item_label }}</a>
+                                {% endif %}
+                            {% enddropdown_button %}
+                        {% else %}
+                            <div class="w-grid">
```
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/workflows/edit_task.html:72]: In other places we're adding a combination of `w-flex` and `w-flex-grow`. I assume this is to ensure the button fills the whole width of the container? I wonder if this is something that could/should be handled in the CSS instead, so we don't have to add the utility classes in different places (with different approaches) which might result in unwanted behaviour if we miss it.

File: wagtail/snippets/templates/wagtailsnippets/snippets/create.html [hunk 12]
```diff
@@ -16,14 +16,8 @@
...
         <footer class="footer">
-            <nav aria-label="{% trans 'Actions' %}">
-                <ul>
-                    <li class="actions actions--primary footer__container">
-                        <div class="dropdown dropup dropdown-button match-width">
-                            {{ action_menu.render_html }}
-                        </div>
-                    </li>
-                </ul>
+            <nav class="actions actions--primary footer__container" aria-label="{% trans 'Actions' %}">
```
Comment (laymonage) [inline wagtail/snippets/templates/wagtailsnippets/snippets/create.html:19]: This doesn't have the `w-flex` and `w-flex-grow` combo and results in incorrect width if there's only one action, e.g. if the user only has "add"/"change" permission (not delete/unpublish), e.g.

image

File: wagtail/admin/templates/wagtailadmin/workflows/edit.html [hunk 13]
```diff
@@ -77,22 +77,28 @@
...
+                                        type="submit"
+                                    >
+                                        {{ view.enable_item_label }}
+                                    </button>
+                                {% elif can_disable %}
+                                    <a href="{{ view.get_delete_url }}" class="button">{{ view.delete_item_label }}</a>
+                                {% endif %}
+                            {% enddropdown_button %}
+                        {% else %}
+                            <div class="w-grid">
```
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/workflows/edit.html:98]: Same here

File: client/scss/components/_footer.scss [hunk 14]
```diff
@@ -1,189 +1,112 @@
...
 
-footer .actions {
-  .button {
-    font-weight: 600;
-    white-space: initial;
+.actions .w-dropdown-button {
+  width: 100%;
+
+  > .button {
+    flex-grow: 1;
```
Comment (laymonage) [inline client/scss/components/_footer.scss:110]: I'd like to know whether there are specific considerations on deciding where such modifications to components should live, i.e. in the component that modifies it (footer in this case) vs. the component that is modified (dropdown-button). Could you clarify?

File: client/scss/overrides/_vendor.tippy.scss [hunk 15]
```diff
@@ -31,9 +31,17 @@
   }
 }
 
+.tippy-box[data-theme='dropdown-button'] {
```
Comment (laymonage) [inline client/scss/overrides/_vendor.tippy.scss:34]: e.g. in this case I guess I kind of understand why this is here and not in `dropdown-button`, but then in `dropdown-button` we also have `[data-tippy-root]`

File: client/scss/components/_dropdown-button.scss (hunk 5 above)
Comment (laymonage) [reply client/scss/components/_dropdown-button.scss:136]: It's probably to prevent accidentally clicking actions that are potentially more destructive, but I don't really see why it's needed as we also have the confirmation page for these actions.

**Review by laymonage — COMMENTED** [review]
This looks good for the most part, thanks! I have a few comments, but they're pretty minor.

**Comment by zerolab** [comment]
question: have you tested the changes with packages that add items to the action bar, like https://github.com/torchbox/wagtail-content-import?

File: client/src/controllers/DropdownController.ts (hunk 9 above)
Comment (thibaudcolas) [reply client/src/controllers/DropdownController.ts:200]: Thank you :)

File: wagtail/admin/templates/wagtailadmin/shared/dropdown/dropdown_button.html [hunk 16]
```diff
@@ -0,0 +1,22 @@
+{% load wagtailadmin_tags i18n %}
+
+{% comment "text/markdown" %}
+    A button with a dropdown menu next to it.
+
+    - `button` (fragment) - the main button
+    - `toggle_icon` (string?) - toggle icon identifier
+    - `toggle_classname` (string?) - additional toggle classes
+    - `children` - Dropdown contents (`a` and `button` elements only)
```
[the remark this answers is not in the document]
Comment (thibaudcolas) [reply wagtail/admin/templates/wagtailadmin/shared/dropdown/dropdown_button.html:9]: ty!

File: wagtail/admin/templates/wagtailadmin/pages/create.html [hunk 17]
```diff
@@ -21,22 +21,13 @@
...
-                            </li>
-
-                            {% block extra_footer_actions %}
-                            {% endblock %}
-                        </ul>
-                    </nav>
-                </li>
-                {% include "wagtailadmin/pages/_unsaved_changes_warning.html" %}
-            </ul>
+        <footer class="footer w-grid md:w-grid-flow-col">
```
[the remark this answers is not in the document]
Comment (thibaudcolas) [reply wagtail/admin/templates/wagtailadmin/pages/create.html:24]: This is just me being inconsistent, thank you! This variant works better with the "unsaved changes warning" on mobile specifically.

File: wagtail/admin/templates/wagtailadmin/workflows/edit_task.html (hunk 11 above)
Comment (thibaudcolas) [reply wagtail/admin/templates/wagtailadmin/workflows/edit_task.html]: Yes, I was keen to further refactor this so we’d have a reusable "footer actions" component, but essentially ran out of time. Will switch to `w-flex-grow` and `w-flex` for now.

File: wagtail/snippets/templates/wagtailsnippets/snippets/create.html (hunk 12 above)
Comment (thibaudcolas) [reply wagtail/snippets/templates/wagtailsnippets/snippets/create.html]: Thank you, I’ll fix that one with `w-grid` at the expense of consistency so we only have to change those two templates.

**Review by laymonage — APPROVED** [review]
Tested on Chrome 120, Firefox 121, and Safari 17.2.1 as well as VoiceOver on Safari on macOS 14.2.1. All works well and the issues have been addressed. Thanks!

I noticed there's a change in layout for the action buttons on the history view, but it's not really breaking and we'll eventually refactor this to use the dropdown component anyway, so I'm happy to leave it for now. 
image

**Comment by laymonage** [comment]
@zerolab I haven't tested it specifically, but the overridden `extra_footer_actions` block is still there. At best, it works as-is albeit with incorrect markup (`li` inside the `div`). At worst, it breaks entirely (which I doubt).

In any case, not sure if that package predated the hook, but there is `construct_page_action_menu` hook that I think if possible should be used instead of replacing the template.

**Comment by zerolab** [comment]
Thank you @laymonage and @thibaudcolas 🙌
